pkg/endpoint/events.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"sync"
+
+	"github.com/cilium/cilium/pkg/eventqueue"
+)
+
+// EndpointRegenerationEvent contains all fields necessary to regenerate an endpoint.
+type EndpointRegenerationEvent struct {
+	owner        Owner
+	regenContext *regenerationContext
+	ep           *Endpoint
+}
+
+// Handle handles the regeneration event for the endpoint.
+func (ev *EndpointRegenerationEvent) Handle(res chan interface{}) {
+	e := ev.ep
+	owner := ev.owner
+	regenContext := ev.regenContext
+
+	err := e.RLockAlive()
+	if err != nil {
+		e.LogDisconnectedMutexAction(err, "before regeneration")
+		res <- &EndpointRegenerationResult{
+			err: err,
+		}
+		return
```

Comment:
Oof, good catch.

Reply:
Acutally, this is OK because `RLockAlive` calls `RLock()` then releases it if the endpoint is disconnecting, and returns an error in that case. So, at this point, the lock actually has been released. 

https://github.com/cilium/cilium/blob/73aafb818b564b673116af008497c28121d1794e/pkg/endpoint/lock.go#L34-L42